### PR TITLE
Fix RBD primary storage host port null error

### PR DESCRIPTION
**Pull request by div8cn**
### Description
Ubuntu 20.04 +KVM +CEPH 15.2

Add RBD main storage through UI, it will fail when there is no host port parameter;
Because when we created the pool, we did not add the port target in the xml

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtDomainXMLParser.java [hunk 1]
```diff
@@ -84,7 +84,11 @@ public boolean parseDomainXML(String domXML) {
                     String authUserName = getAttrValue("auth", "username", disk);
                     String poolUuid = getAttrValue("secret", "uuid", disk);
                     String host = getAttrValue("host", "name", disk);
-                    int port = Integer.parseInt(getAttrValue("host", "port", disk));
+                    int port = 0;
+                    String _xmlPort = getAttrValue("host", "port", disk);
+                    if ( ! _xmlPort.isEmpty()) {
```
Comment (rhtyd) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtDomainXMLParser.java:89]: Should we do both null and empty check?

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtStoragePoolXMLParser.java [hunk 2]
```diff
@@ -55,7 +55,11 @@ public LibvirtStoragePoolDef parseStoragePoolXML(String poolXML) {
             String format = getAttrValue("format", "type", source);
 
             if (type.equalsIgnoreCase("rbd")) {
-                int port = Integer.parseInt(getAttrValue("host", "port", source));
+                int port = 0;
+                String _xmlPort = getAttrValue("host", "port", source);
+                if ( ! _xmlPort.isEmpty()) {
```
Comment (rhtyd) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtStoragePoolXMLParser.java:60]: Should we do both null and empty check?

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtDomainXMLParser.java (hunk 1 above)
Comment (wido) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtDomainXMLParser.java:89]: I would do that just for the sake of security

Comment (DaanHoogland) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtDomainXMLParser.java:89]: there is a lot of npe danger in `getAttrValue()`, none of teh parameters are checked. (meaning yes please)

**Comment by rhtyd** [comment]
@div8cn with you comment https://github.com/apache/cloudstack/pull/4568#issuecomment-754321932 do we need this PR?

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtDomainXMLParser.java (hunk 1 above)
Comment (DaanHoogland) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtDomainXMLParser.java:89]: I did an investigation of the dom code used in `getAttrValue()` and it returns an empty string if the element `"host"` does not have an attribute named `"port"`, so _xmlPort will not be null.

**Comment by div8cn** [comment]
@rhtyd I think this PR is necessary,

The best practice is to let libvirt decide to use port 3300 or 6789 to communicate with CEPH.

Therefore, it is best not to enter any port when adding mon;

@wido What do you think?

**Comment by wido** [comment]
I agree @div8cn , this PR is needed.

In the early days of Ceph you needed to specify the port. Now we should have libvirt/qemu/librados/librbd handle this. We shouldn't interfere.

People running Ceph monitors on alternative ports are rare, very rare. I have never encountered them recently.

So yes, just a hostname should be sufficient without a port number.

**Comment by rhtyd** [comment]
@wido @DaanHoogland @GabrielBrascher @weizhouapache  should this be treated as a blocker then, or good to have?
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtDomainXMLParser.java (hunk 1 above)
Comment (GabrielBrascher) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtDomainXMLParser.java:89]: 1. @div8cn can you please change the variable name from `_xmlPort` to `xmlPort`?

According to [Java conventions](https://docs.oracle.com/javase/tutorial/java/nutsandbolts/variables.html), 

> while it's technically legal to begin your variable's name with "_", this practice is discouraged

2. Regarding the String validation. Is it possible to have a `tagNode` of length 0?

`getAttrValue()` has this piece of code:
```
        NodeList tagNode = eElement.getElementsByTagName(tag);
        if (tagNode.getLength() == 0) {
            return null;
        }
```

I would recommend, just for safety, to use something as `org.apache.commons.lang.StringUtils`; it provides String validations such as `StringUtils.isNotBlank(Str)`.

`isNotBlank` returns true only if not null, not empty, and not blank.

```
String xmlPort = getAttrValue("host", "port", disk);
if (StringUtils.isNotBlank(xmlPort)) {
     ....
}
```

**Comment by GabrielBrascher** [comment]
@rhtyd it would be great to have this one on board, but I don't see reason in blocking a release because of it.

Any comments, @wido @DaanHoogland @div8cn @weizhouapache?

**Review by GabrielBrascher — APPROVED** [review]
LGTM based on code review.

**Review by rhtyd — APPROVED** [review]

**Pull request merged**